File: crates/dewasm-backend-perl/units/wasi/wasi_filetype.pl
# Map a stat mode word to a WASI filetype tag (S_IFMT bits, avoiding a
# POSIX-constant dependency).
sub wasi_filetype {
    my ($self, $mode) = @_;
    my $fmt = $mode & 0170000;
    return 3 if $fmt == 0040000;  # directory
    return 2 if $fmt == 0020000;  # character device
    return 1 if $fmt == 0060000;  # block device
    return 7 if $fmt == 0120000;  # symbolic link
    return 6 if $fmt == 0140000;  # socket (stream)
    return 4 if $fmt == 0100000;  # regular file
    return 0;  # unknown
}

File: crates/dewasm-backend-python/units/wasi/wasi_filetype.py
# Map a stat result's st_mode to a WASI filetype tag (S_IFMT bits, avoiding a dependency on the `stat` module).
def wasi_filetype(self, st):
    fmt = st.st_mode & 0o170000
    if fmt == 0o040000:
        return 3  # directory
    if fmt == 0o020000:
        return 2  # character device
    if fmt == 0o060000:
        return 1  # block device
    if fmt == 0o120000:
        return 7  # symbolic link
    if fmt == 0o140000:
        return 6  # socket (stream)
    if fmt == 0o100000:
        return 4  # regular file
    return 0  # unknown
